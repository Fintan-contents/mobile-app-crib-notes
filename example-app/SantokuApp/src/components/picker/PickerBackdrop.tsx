import {composePressableStyles} from 'framework/utilities';
import React, {useMemo} from 'react';
import {Modal as RNModal, ModalProps, Omit, Pressable, PressableProps, StyleSheet, ViewProps} from 'react-native';
import Reanimated, {BaseAnimationBuilder, FadeIn, FadeOut, Keyframe} from 'react-native-reanimated';

import {usePickerBackdropUseCase} from './usePickerBackdropUseCase';

export const DEFAULT_COLOR = 'rgba(0,0,0,0.4)';
export const DEFAULT_FADE_IN_DURATION = 300;
export const DEFAULT_FADE_OUT_DURATION = 150;
export const DEFAULT_ENTERING = FadeIn.duration(DEFAULT_FADE_IN_DURATION);
export const DEFAULT_EXITING = FadeOut.duration(DEFAULT_FADE_OUT_DURATION);

export type PickerBackdropProps = Omit<Reanimated.AnimateProps<ViewProps>, 'entering' | 'exiting'> & {
  isVisible: boolean;
  onPress?: () => unknown;
  /**
   * iOSの場合、アニメーションが終わった後に呼び出されます。
   * Androidの場合、アニメーションが始まった時に呼び出されます。
   */
  enteringCallback?: (finished: boolean) => unknown;
  exitingCallback?: (finished: boolean) => unknown;
  pressableProps?: PressableProps;
  modalProps?: ModalProps;
  /**
   * enteringに指定したAnimationBuilderなどでwithCallbackを指定しても、このコンポーネントの中で上書きしているため実行できません。
   * withCallbackで実行する関数は、enteringCallbackで指定してください。
   */
  entering?: BaseAnimationBuilder | Keyframe;
  /**
   * exitingに指定したAnimationBuilderなどでwithCallbackを指定しても、このコンポーネントの中で上書きしているため実行できません。
   * withCallbackで実行する関数は、exitingCallbackで指定してください。
   */
  exiting?: BaseAnimationBuilder | Keyframe;
};

export const PickerBackdrop: React.FC<PickerBackdropProps> = ({
  isVisible,
  onPress,
  entering = DEFAULT_ENTERING,
  exiting = DEFAULT_EXITING,
  enteringCallback,
  exitingCallback,
  pressableProps: {style: pressableStyle, ...pressableProps} = {},
  modalProps: {style: modalStyle, ...modalProps} = {},
  /**
   * このコンポーネントでは、ReanimatedのEntering/Exitingを使用してアニメーションを実現しています。
   * Entering/Exitingを使用した場合は、opacityを設定しても反映されません。
   * backgroundColorにrgbaで指定してください。（e.g. backgroundColor: rgba(0,0,0,0.4)）
   */
  style,
  children,
  ...animatedViewProps
}) => {
  const {isModalVisible, composedEnteringCallback, composedExitingCallback} = usePickerBackdropUseCase({
    isVisible,
    enteringCallback,
    exitingCallback,
  });

  const composedPressableStyles = useMemo(
    () => composePressableStyles([styles.pressable, pressableStyle]),
    [pressableStyle],
  );

  return !isModalVisible ? null : (
    <RNModal
      visible
      statusBarTranslucent
      animationType="none"
      transparent
      // 戻るボタンが押されたとき（onRequestClose）は、背景がタップされたときと同じ振る舞いになるようにしておく。
      onRequestClose={onPress}
      style={modalStyle}
      {...modalProps}>
      {isVisible && (
        <Pressable onPress={onPress} style={composedPressableStyles} {...pressableProps}>
          <Reanimated.View
            entering={entering.withCallback(composedEnteringCallback)}
            exiting={exiting.withCallback(composedExitingCallback)}
            style={StyleSheet.flatten([styles.backdrop, style])}
            {...animatedViewProps}
          />
        </Pressable>
      )}
      {children}
    </RNModal>
  );
};

const styles = StyleSheet.create({
  pressable: {
    ...StyleSheet.absoluteFillObject,
    display: 'flex',
  },
  backdrop: {
    flex: 1,
    backgroundColor: DEFAULT_COLOR,
  },
});
