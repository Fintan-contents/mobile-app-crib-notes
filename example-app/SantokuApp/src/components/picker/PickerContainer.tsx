import {useWorkletCallback} from 'framework/utilities/useWorkletCallback';
import React from 'react';
import {StyleSheet, View, ViewProps} from 'react-native';
import Reanimated, {BaseAnimationBuilder, Keyframe, SlideInDown, SlideOutDown} from 'react-native-reanimated';

export type PickerContainerProps = Omit<Reanimated.AnimateProps<ViewProps>, 'entering' | 'exiting'> & {
  isVisible: boolean;
  /**
   * iOSの場合、アニメーションが終わった後に呼び出されます。
   * Androidの場合、アニメーションが始まった時に呼び出されます。
   */
  enteringCallback?: (finished: boolean) => unknown;
  exitingCallback?: (finished: boolean) => unknown;
  /**
   * enteringに指定したAnimationBuilderなどでwithCallbackを指定しても、本コンポーネント内で上書きしているため実行できません。
   * withCallbackで実行する関数は、enteringCallbackで指定してください。
   */
  entering?: BaseAnimationBuilder | Keyframe;
  /**
   * exitingに指定したAnimationBuilderなどでwithCallbackを指定しても、本コンポーネント内で上書きしているため実行できません。
   * withCallbackで実行する関数は、exitingCallbackで指定してください。
   */
  exiting?: BaseAnimationBuilder | Keyframe;
};

export const PICKER_CONTAINER_DEFAULT_SLIDE_IN_DURATION = 300;
export const PICKER_CONTAINER_DEFAULT_SLIDE_OUT_DURATION = 300;
export const PICKER_CONTAINER_DEFAULT_ENTERING = SlideInDown.duration(PICKER_CONTAINER_DEFAULT_SLIDE_IN_DURATION);
export const PICKER_CONTAINER_DEFAULT_EXITING = SlideOutDown.duration(PICKER_CONTAINER_DEFAULT_SLIDE_OUT_DURATION);

export const PickerContainer: React.FC<PickerContainerProps> = ({
  isVisible,
  entering = PICKER_CONTAINER_DEFAULT_ENTERING,
  exiting = PICKER_CONTAINER_DEFAULT_EXITING,
  enteringCallback,
  exitingCallback,
  style,
  children,
  ...animatedViewProps
}) => {
  const composedEnteringCallback = useWorkletCallback(enteringCallback);
  const composedExitingCallback = useWorkletCallback(exitingCallback);

  return (
    <>
      {isVisible && (
        <View style={styles.container} pointerEvents="box-none">
          <Reanimated.View
            style={style}
            entering={entering.withCallback(composedEnteringCallback)}
            exiting={exiting.withCallback(composedExitingCallback)}
            pointerEvents="box-none"
            {...animatedViewProps}>
            {children}
          </Reanimated.View>
        </View>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
  },
});
