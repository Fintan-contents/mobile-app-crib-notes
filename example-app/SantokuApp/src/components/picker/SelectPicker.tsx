import {PickerProps} from '@react-native-picker/picker';
import {log} from 'framework/logging';
import React from 'react';
import {Platform, StyleProp, TextStyle, TextInputProps as RNETextInputProps, ViewProps} from 'react-native';

import {DefaultPickerAccessoryProps} from './DefaultPickerAccessory';
import {PickerBackdropProps} from './PickerBackdrop';
import {PickerContainerProps} from './PickerContainer';

export type Item<T> = {
  label: string;
  value: T;
  inputLabel?: string;
  key?: React.Key;
  color?: string;
  fontFamily?: string;
  /**
   * アイテムに指定するスタイル
   * 以下のプロパティのみ使用できます。
   * - 'color'
   * - 'backgroundColor'
   * - 'fontSize'
   * - 'fontFamily'
   *
   * @platform android
   */
  style?: StyleProp<Pick<TextStyle, 'color' | 'backgroundColor' | 'fontSize' | 'fontFamily'>>;
};

type TextInputProps = Omit<RNETextInputProps, 'value' | 'editable'>;
export type SelectPickerProps<ItemT> = {
  /**
   * SelectPickerに表示するアイテム
   */
  items: Item<ItemT>[];
  /**
   * 選択されたアイテムのKey
   */
  selectedItemKey?: React.Key;
  /**
   * アイテムが選択された場合に呼び出される関数
   */
  onSelectedItemChange?: (selectedItem?: Item<ItemT>) => void;
  /**
   * PickerBackdropをタップして閉じた場合に呼び出される関数
   *
   * @platform ios
   */
  onDismiss?: (selectedItem?: Item<ItemT>) => void;
  /**
   * DeleteLabelがタップされた場合に呼び出される関数
   * タップ後、SelectPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  onDelete?: (selectedItem?: Item<ItemT>) => void;
  /**
   * CancelLabelがタップされた場合に呼び出される関数
   * タップ後、SelectPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  onCancel?: (selectedItem?: Item<ItemT>) => void;
  /**
   * [iOS]
   * DoneLabelがタップされた場合に呼び出される関数
   * タップ後、SelectPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * [Android]
   * アイテムが選択された場合に呼び出される関数
   */
  onDone?: (selectedItem?: Item<ItemT>) => void;
  /**
   * プレースホルダ
   */
  placeholder?: string;
  /**
   * ヘッダコンポーネント
   * 指定しない場合は、PickerAccessoryがデフォルトで表示されます。
   *
   * @platform ios
   */
  pickerAccessoryComponent?: React.ReactNode;
  /**
   * Pickerコンポーネント
   * 指定しない場合は、SelectPickerItemsがデフォルトで表示されます。
   */
  pickerItemsComponent?: React.ReactNode;
  /**
   * 選択されたアイテムを表示するテキストコンポーネント
   */
  textInputComponent?: React.ReactNode;
  /**
   * Keyを作成する関数
   */
  keyExtractor?: ((item: ItemT, index: number) => React.Key) | undefined;
  /**
   * 選択されたアイテムを表示するテキストコンポーネントのProps
   * なお、textInputComponentを指定した場合は使用されません。
   */
  textInputProps?: TextInputProps;
  /**
   * Pickerコンポーネントを囲むContainerのProps
   *
   * @platform ios
   */
  pickerItemsContainerProps?: ViewProps;
  /**
   * PickerのProps
   * なお、pickerItemsComponentを指定した場合は使用されません。
   */
  pickerProps?: Omit<PickerProps<ItemT>, 'selectedValue' | 'onValueChange'>;
  /**
   * PickerBackdropのProps
   *
   * @platform ios
   */
  pickerBackdropProps?: Omit<PickerBackdropProps, 'isVisible' | 'onPress'>;
  /**
   * PickerContainerのProps
   *
   * @platform ios
   */
  pickerContainerProps?: Omit<PickerContainerProps, 'isVisible'>;
  /**
   * PickerAccessoryのProps
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  pickerAccessoryProps?: Omit<DefaultPickerAccessoryProps, 'onDelete' | 'onCancel' | 'onDone'>;
};

export const SelectPicker = <ItemT extends unknown>(props: SelectPickerProps<ItemT>) => {
  React.useEffect(() => {
    log.warn(`SelectPicker is not supported on: ${Platform.OS}`);
  }, []);
  return null;
};
