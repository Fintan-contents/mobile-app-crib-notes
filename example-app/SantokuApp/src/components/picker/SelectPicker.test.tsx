import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {TextInputProps, View, ViewProps} from 'react-native';

import {PickerBackdropProps} from './PickerBackdrop';
import {PickerContainerProps} from './PickerContainer';
import {Item, SelectPicker} from './SelectPicker';
import {SelectPickerItemsProps} from './SelectPickerItems';

describe('SelectPicker only with required props', () => {
  it('renders successfully if invisible', () => {
    const items = [
      {value: '1', label: 'test1'},
      {value: '2', label: 'test2'},
    ];
    const sut = render(
      <SelectPicker
        items={items}
        selectedItemKey="1"
        textInputProps={{testID: 'textInput'}}
        pickerBackdropProps={{testID: 'pickerBackdrop'}}
        pickerContainerProps={{testID: 'pickerContainer'}}
        pickerAccessoryProps={{containerProps: {testID: 'pickerAccessory'}}}
        pickerItemsContainerProps={{testID: 'pickerItemsContainer'}}
      />,
    );

    expect(sut).toMatchSnapshot('SelectPicker if invisible.');

    const pickerBackdrop = sut.queryByTestId('pickerBackdrop');
    const pickerContainer = sut.queryByTestId('pickerContainer');
    const pickerAccessory = sut.queryByTestId('pickerAccessory');
    const pickerItemsContainer = sut.queryByTestId('pickerItemsContainer');
    const textInput = sut.queryByTestId('textInput');
    expect(pickerBackdrop).toBeNull();
    expect(pickerContainer).toBeNull();
    expect(pickerAccessory).toBeNull();
    expect(pickerItemsContainer).toBeNull();
    expect(textInput).not.toBeNull();
  });

  it('renders successfully if visible', () => {
    const items = [
      {value: '1', label: 'test1'},
      {value: '2', label: 'test2'},
    ];
    const sut = render(
      <SelectPicker
        items={items}
        selectedItemKey="1"
        textInputProps={{testID: 'textInput'}}
        pickerBackdropProps={{testID: 'pickerBackdrop'}}
        pickerContainerProps={{testID: 'pickerContainer'}}
        pickerAccessoryProps={{containerProps: {testID: 'pickerAccessory'}}}
        pickerItemsContainerProps={{testID: 'pickerItemsContainer'}}
      />,
    );
    const pressableContainer = sut.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    expect(sut).toMatchSnapshot('SelectPicker if visible.');

    const pickerBackdrop = sut.queryByTestId('pickerBackdrop');
    const pickerContainer = sut.queryByTestId('pickerContainer');
    const pickerAccessory = sut.queryByTestId('pickerAccessory');
    const pickerItemsContainer = sut.queryByTestId('pickerItemsContainer');
    const textInput = sut.queryByTestId('textInput');
    expect(pickerBackdrop).not.toBeNull();
    expect(pickerContainer).not.toBeNull();
    expect(pickerAccessory).not.toBeNull();
    expect(pickerItemsContainer).not.toBeNull();
    expect(textInput).not.toBeNull();
  });
});

describe('SelectPicker with all props', () => {
  it('should be applied all properly with default xxx component', () => {
    const items = [
      {value: '1', label: 'test1'},
      {value: '2', label: 'test2'},
    ];
    const onSelectedItemChange = jest.fn();
    const onDismiss = jest.fn();
    const onDelete = jest.fn();
    const onCancel = jest.fn();
    const onDone = jest.fn();
    const keyExtractor = jest.fn((item: Item<string>, index: number) => `${String(item.key ?? item.value)}-${index}`);
    const sut = render(
      <SelectPicker
        items={items}
        selectedItemKey="1"
        onSelectedItemChange={onSelectedItemChange}
        onDismiss={onDismiss}
        placeholder="please select..."
        keyExtractor={keyExtractor}
        textInputProps={{style: {color: 'red'}, testID: 'textInput'}}
        pickerItemsContainerProps={{pointerEvents: 'none', testID: 'pickerItemsContainer'}}
        pickerProps={{numberOfLines: 5, testID: 'picker'}}
        pickerBackdropProps={{
          style: {backgroundColor: 'green', borderColor: 'red'},
          modalProps: {testID: 'modal'},
          testID: 'pickerBackdrop',
        }}
        pickerContainerProps={{style: {backgroundColor: 'yellow', borderColor: 'black'}, testID: 'pickerContainer'}}
        pickerAccessoryProps={{
          containerProps: {style: {backgroundColor: 'blue'}, testID: 'pickerAccessory'},
          deleteLabel: 'delete',
          cancelLabel: 'cancel',
          doneLabel: 'done',
          deleteTouchableContainerProps: {testID: 'deleteLink'},
          cancelTouchableContainerProps: {testID: 'cancelLink'},
          doneTouchableContainerProps: {testID: 'doneLink'},
        }}
        onDelete={onDelete}
        onCancel={onCancel}
        onDone={onDone}
      />,
    );
    const pressableContainer = sut.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    expect(sut).toMatchSnapshot('SelectPicker all properly with default xxx component.');

    // assert pickerContainer
    const pickerContainer = sut.getByTestId('pickerContainer');
    const pickerContainerProps = pickerContainer.props as PickerContainerProps;
    expect(pickerContainerProps.style).toEqual({backgroundColor: 'yellow', borderColor: 'black'});

    // assert pickerItemsContainer
    const pickerItemsContainer = sut.getByTestId('pickerItemsContainer');
    const pickerItemsContainerProps = pickerItemsContainer.props as ViewProps;
    expect(pickerItemsContainerProps.pointerEvents).toBe('none');

    // assert PickerItems
    const picker = sut.getByTestId('picker');
    const pickerProps = picker.props as SelectPickerItemsProps<string>;
    expect(pickerProps.numberOfLines).toBe(5);
    expect(pickerProps.items.length).toBe(2);

    // assert textInput
    const textInput = sut.getByTestId('textInput');
    const textInputProps = textInput.props as TextInputProps;
    expect(textInputProps.style).toEqual({color: 'red'});
    expect(keyExtractor).toHaveBeenNthCalledWith(1, {value: '1', label: 'test1'}, 0);
    expect(keyExtractor).toHaveBeenNthCalledWith(2, {value: '2', label: 'test2'}, 1);

    // assert pickerBackdrop
    const pickerBackdrop = sut.getByTestId('pickerBackdrop');
    const modal = sut.getByTestId('modal');
    const pickerBackdropProps = pickerBackdrop.props as PickerBackdropProps;
    fireEvent(modal, 'onRequestClose');
    expect(pickerBackdropProps.style).toEqual({flex: 1, backgroundColor: 'green', borderColor: 'red'});
    expect(onDismiss).toHaveBeenCalledTimes(1);

    fireEvent.press(pressableContainer);

    // assert pickerAccessory
    const pickerAccessory = sut.getByTestId('pickerAccessory');
    const pickerAccessoryProps = pickerAccessory.props as ViewProps;
    expect(pickerAccessoryProps.style).toEqual({
      backgroundColor: 'blue',
      flexDirection: 'row',
      justifyContent: 'flex-end',
      paddingHorizontal: 10,
      paddingVertical: 8,
    });
    const deleteLink = sut.getByTestId('deleteLink');
    const deleteText = sut.queryByText('delete');
    expect(deleteText).not.toBeNull();
    fireEvent.press(deleteLink);
    expect(onDelete).toHaveBeenCalledTimes(1);
    fireEvent.press(pressableContainer);
    const cancelLink = sut.getByTestId('cancelLink');
    const cancelText = sut.queryByText('cancel');
    expect(cancelText).not.toBeNull();
    fireEvent.press(cancelLink);
    expect(onCancel).toHaveBeenCalledTimes(1);
    fireEvent.press(pressableContainer);
    const doneLink = sut.getByTestId('doneLink');
    const doneText = sut.queryByText('done');
    expect(doneText).not.toBeNull();
    fireEvent.press(doneLink);
    expect(onDone).toHaveBeenCalledTimes(1);
  });

  it('should be applied all properly with custom xxx component', () => {
    const CustomPickerAccessory = <View testID="customPickerAccessory" />;
    const CustomPicker = <View testID="customPicker" />;
    const CustomTextInput = <View testID="customTextInput" />;
    const items = [
      {value: '1', label: 'test1'},
      {value: '2', label: 'test2'},
    ];
    const onSelectedItemChange = jest.fn();
    const onDismiss = jest.fn();
    const sut = render(
      <SelectPicker
        items={items}
        selectedItemKey="1"
        onSelectedItemChange={onSelectedItemChange}
        onDismiss={onDismiss}
        placeholder="please select..."
        textInputProps={{testID: 'defaultTextInput'}}
        pickerItemsContainerProps={{pointerEvents: 'none', testID: 'pickerItemsContainer'}}
        pickerProps={{testID: 'defaultPicker'}}
        pickerBackdropProps={{
          style: {backgroundColor: 'green', borderColor: 'red'},
          modalProps: {testID: 'modal'},
          testID: 'pickerBackdrop',
        }}
        pickerContainerProps={{style: {backgroundColor: 'yellow', borderColor: 'black'}, testID: 'pickerContainer'}}
        pickerAccessoryProps={{containerProps: {testID: 'defaultPickerAccessory'}}}
        pickerAccessoryComponent={CustomPickerAccessory}
        pickerItemsComponent={CustomPicker}
        textInputComponent={CustomTextInput}
      />,
    );
    const pressableContainer = sut.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    expect(sut).toMatchSnapshot('SelectPicker all properly with custom xxx component.');

    // assert pickerBackdrop
    const pickerBackdrop = sut.getByTestId('pickerBackdrop');
    const modal = sut.getByTestId('modal');
    const pickerBackdropProps = pickerBackdrop.props as PickerBackdropProps;
    fireEvent(modal, 'onRequestClose');
    expect(pickerBackdropProps.style).toEqual({flex: 1, backgroundColor: 'green', borderColor: 'red'});
    expect(onDismiss).toHaveBeenCalledTimes(1);

    fireEvent.press(pressableContainer);

    // assert pickerContainer
    const pickerContainer = sut.getByTestId('pickerContainer');
    const pickerContainerProps = pickerContainer.props as PickerContainerProps;
    expect(pickerContainerProps.style).toEqual({backgroundColor: 'yellow', borderColor: 'black'});

    // assert pickerAccessory
    const defaultPickerAccessory = sut.queryByTestId('defaultPickerAccessory');
    const customPickerAccessory = sut.queryByTestId('customPickerAccessory');
    expect(defaultPickerAccessory).toBeNull();
    expect(customPickerAccessory).not.toBeNull();

    // assert pickerItemsContainer
    const pickerItemsContainer = sut.getByTestId('pickerItemsContainer');
    const pickerItemsContainerProps = pickerItemsContainer.props as ViewProps;
    expect(pickerItemsContainerProps.pointerEvents).toBe('none');

    // assert Picker
    const defaultPicker = sut.queryByTestId('defaultPicker');
    const customPicker = sut.queryByTestId('customPicker');
    expect(defaultPicker).toBeNull();
    expect(customPicker).not.toBeNull();

    // assert textInput
    const defaultTextInput = sut.queryByTestId('defaultTextInput');
    const customTextInput = sut.queryByTestId('customTextInput');
    expect(defaultTextInput).toBeNull();
    expect(customTextInput).not.toBeNull();
  });
});
