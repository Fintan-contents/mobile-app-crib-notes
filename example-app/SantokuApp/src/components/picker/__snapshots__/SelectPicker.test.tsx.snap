// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectPicker only with required props renders successfully if invisible: SelectPicker if invisible. 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="pressableContainer"
>
  <View
    pointerEvents="box-only"
  >
    <TextInput
      editable={false}
      testID="textInput"
      value="test1"
    />
  </View>
</View>
`;

exports[`SelectPicker only with required props renders successfully if visible: SelectPicker if visible. 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="pickerBackdrop"
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateY": 1000,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "transform": Array [
              Object {
                "translateY": 1000,
              },
            ],
          }
        }
        testID="pickerContainer"
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "paddingHorizontal": 10,
              "paddingVertical": 8,
            }
          }
          testID="pickerAccessory"
        />
        <View
          testID="pickerItemsContainer"
        >
          <View>
            <RNCPicker
              items={
                Array [
                  Object {
                    "label": "test1",
                    "testID": undefined,
                    "textColor": undefined,
                    "value": "1",
                  },
                  Object {
                    "label": "test2",
                    "testID": undefined,
                    "textColor": undefined,
                    "value": "2",
                  },
                ]
              }
              numberOfLines={1}
              onChange={[Function]}
              selectedIndex={0}
              style={
                Array [
                  Object {
                    "height": 216,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </Modal>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressableContainer"
  >
    <View
      pointerEvents="box-only"
    >
      <TextInput
        editable={false}
        testID="textInput"
        value="test1"
      />
    </View>
  </View>,
]
`;

exports[`SelectPicker with all props should be applied all properly with custom xxx component: SelectPicker all properly with custom xxx component. 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    testID="pickerBackdropModal"
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "red",
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="pickerBackdropPressable"
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateY": 1000,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "yellow",
            "borderColor": "black",
            "transform": Array [
              Object {
                "translateY": 1000,
              },
            ],
          }
        }
        testID="pickerContainer"
      >
        <View
          testID="customPickerAccessory"
        />
        <View
          pointerEvents="none"
          testID="pickerItemsContainer"
        >
          <View
            testID="customPicker"
          />
        </View>
      </View>
    </View>
  </Modal>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressableContainer"
  >
    <View
      pointerEvents="box-only"
    >
      <View
        testID="customTextInput"
      />
    </View>
  </View>,
]
`;

exports[`SelectPicker with all props should be applied all properly with default xxx component: SelectPicker all properly with default xxx component. 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    testID="pickerBackdropModal"
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "red",
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="pickerBackdropPressable"
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateY": 1000,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "yellow",
            "borderColor": "black",
            "transform": Array [
              Object {
                "translateY": 1000,
              },
            ],
          }
        }
        testID="pickerContainer"
      >
        <View
          style={
            Object {
              "backgroundColor": "blue",
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "paddingHorizontal": 10,
              "paddingVertical": 8,
            }
          }
          testID="pickerAccessory"
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="deleteLink"
          >
            <Text
              style={
                Object {
                  "color": "#d9545e",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              delete
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="cancelLink"
          >
            <Text
              style={
                Object {
                  "color": "#4577CC",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              cancel
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="doneLink"
          >
            <Text
              style={
                Object {
                  "color": "#4577CC",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              done
            </Text>
          </View>
        </View>
        <View
          pointerEvents="none"
          testID="pickerItemsContainer"
        >
          <View>
            <RNCPicker
              items={
                Array [
                  Object {
                    "label": "test1",
                    "testID": undefined,
                    "textColor": undefined,
                    "value": "1",
                  },
                  Object {
                    "label": "test2",
                    "testID": undefined,
                    "textColor": undefined,
                    "value": "2",
                  },
                ]
              }
              numberOfLines={5}
              onChange={[Function]}
              selectedIndex={0}
              style={
                Array [
                  Object {
                    "height": 216,
                  },
                  undefined,
                ]
              }
              testID="picker"
            />
          </View>
        </View>
      </View>
    </View>
  </Modal>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressableContainer"
  >
    <View
      pointerEvents="box-only"
    >
      <TextInput
        editable={false}
        placeholder="please select..."
        style={
          Object {
            "color": "red",
          }
        }
        testID="textInput"
        value="test1"
      />
    </View>
  </View>,
]
`;
